apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
-   * negative acknowledgement. If paper already parked, does nothing and returns
-   * positive acknowledgement. When parked, parkSensor should be true.
-   */
-  async parkPaper(): Promise<void> {
-    debug('+parkPaper');
-    this.status = 'parking_paper';
-    this.lastActionInitiatedTime = Date.now();
-    await this.driver.parkPaper();
-    debug('-parkPaper');
-  }
+function paperHandlerStatusToEvent(
+  paperHandlerStatus: PaperHandlerStatus
+): PaperHandlerStatusEvent {
+  let event: PaperHandlerStatusEvent = { type: 'NO_PAPER_IN_FRONT' };
 
-  /**
-   * Moves paper to the front for voter to see, but hangs on to the paper.
-   * Equivalent to "reject hold." How do we differentiate the present paper
-   * state from the state where paper has not been picked up yet?
-   */
-  async presentPaper(): Promise<void> {
-    await this.driver.presentPaper();
+  if (isPaperInScanner(paperHandlerStatus)) {
+    if (isPaperInOutput(paperHandlerStatus)) {
+      event = { type: 'BALLOT_PRINTED' };
+    } else {
+      event = { type: 'PAPER_INSIDE' };
+    }
+  } else if (isPaperReadyToLoad(paperHandlerStatus)) {
+    event = { type: 'PAPER_READY_TO_LOAD' };
   }
+  debug(`Emitting event ${event.type}`);
+  return event;
+}
 
-  /**
-   * Ejects out the back. Can eject from loaded or parked state. If there is
-   * no paper to eject, handler will do nothing and return positive acknowledgement.
-   */
-  async eject(): Promise<void> {
-    this.status = 'ejecting';
-    await this.driver.ejectBallot();
-  }
+/**
+ * Builds an observable that polls paper status and emits state machine events.
+ * Notes:
+ * Why Observable? This section from the xstate docs matches our use case closely:
+ * https://xstate.js.org/docs/guides/communication.html#invoking-observables
+ * "Observables can be invoked, which is expected to send events (strings or objects) to the parent machine,
+ * yet not receive events (uni-directional). An observable invocation is a function that takes context and
+ * event as arguments and returns an observable stream of events."
+ *
+ *
+ */
+function buildPaperStatusObservable() {
+  return ({ driver, pollingIntervalMs }: Context) => {
+    // `timer` returns an Observable that emits values with `pollingInterval` delay between each event
+    return (
+      timer(0, pollingIntervalMs)
+        // `pipe` forwards the value from the previous function to the next unary function ie. switchMap.
+        // In this case there is no value. The combination of timer(...).pipe() is so we can execute the
+        // function supplied to `switchMap` on the specified interval.
+        .pipe(
+          // `switchMap` returns an Observable that emits events.
+          // Why do we use `switchMap`? Is it because `switchMap` only emits items from the most recent
+          // inner Observable ie. it will only emit the latest PaperHandlerStatusEvent?
+          switchMap(async () => {
```

Comment:
@jonahkagan I think I have a decent idea of how this implementation of the status poller works except for our use of `swichMap` here. Do you remember why we use it?

Reply:
I don't remember, but maybe it has to do with making the timeout work?